process/block/preprocess/transactionsV2.go

```diff
@@ -0,0 +1,416 @@
+		go txs.notifyTransactionProviderIfNeeded()
+	}()
+
+	for shardID := uint32(0); shardID < txs.shardCoordinator.NumberOfShards(); shardID++ {
+		mapMiniBlocks[shardID] = txs.createEmptyMiniBlock(txs.shardCoordinator.SelfId(), shardID, block.TxBlock)
+	}
+
+	mapMiniBlocks[core.MetachainShardId] = txs.createEmptyMiniBlock(txs.shardCoordinator.SelfId(), core.MetachainShardId, block.TxBlock)
+
+	for index := range sortedTxs {
+		if !haveTime() {
+			log.Debug("time is out in createAndProcessMiniBlocksFromMeV2")
+			break
+		}
+
+		tx, ok := sortedTxs[index].Tx.(*transaction.Transaction)
+		if !ok {
+			log.Debug("wrong type assertion",
+				"hash", sortedTxs[index].TxHash,
+				"sender shard", sortedTxs[index].SenderShardID,
+				"receiver shard", sortedTxs[index].ReceiverShardID)
+			continue
+		}
+
+		txHash := sortedTxs[index].TxHash
+		senderShardID := sortedTxs[index].SenderShardID
+		receiverShardID := sortedTxs[index].ReceiverShardID
+
+		miniBlock, ok := mapMiniBlocks[receiverShardID]
+		if !ok {
+			log.Debug("miniblock is not created", "shard", receiverShardID)
+			continue
+		}
+
+		isReceiverSmartContractAddress := core.IsSmartContractAddress(tx.RcvAddr)
+		firstMiniBlockSplitForReceiverShardFound := (isReceiverSmartContractAddress && mapScsForShard[receiverShardID] == 0 && mapTxsForShard[receiverShardID] > 0) ||
+			(!isReceiverSmartContractAddress && mapTxsForShard[receiverShardID] == 0 && mapScsForShard[receiverShardID] > 0)
+
+		numNewMiniBlocks := 0
+		if len(miniBlock.TxHashes) == 0 || firstCrossShardScCallOrSpecialTxFound {
+			numNewMiniBlocks = 1
+		}
+		numNewTxs := 1
+
+		isCrossShardScCallOrSpecialTx := receiverShardID != txs.shardCoordinator.SelfId() &&
+			(isReceiverSmartContractAddress || len(tx.RcvUserName) > 0)
+		if isCrossShardScCallOrSpecialTx {
+			if !firstCrossShardScCallOrSpecialTxFound {
+				numNewMiniBlocks++
+			}
+			numNewTxs += core.AdditionalScrForEachScCallOrSpecialTx
+		}
+
+		if isMaxBlockSizeReached(numNewMiniBlocks, numNewTxs) {
+			log.Debug("max txs accepted in one block is reached",
+				"num txs added", numTxsAdded,
+				"total txs", len(sortedTxs))
+			break
+		}
+
+		if isShardStuck != nil && isShardStuck(receiverShardID) {
+			log.Trace("shard is stuck", "shard", receiverShardID)
+			continue
+		}
+
+		if len(senderAddressToSkip) > 0 {
+			if bytes.Equal(senderAddressToSkip, tx.GetSndAddr()) {
+				numTxsSkipped++
+				continue
+			}
+		}
+
+		txMaxTotalCost := big.NewInt(0)
+		isAddressSet := txs.balanceComputation.IsAddressSet(tx.GetSndAddr())
+		if isAddressSet {
+			txMaxTotalCost = txs.getTxMaxTotalCost(tx)
+		}
+
+		if isAddressSet {
+			addressHasEnoughBalance := txs.balanceComputation.AddressHasEnoughBalance(tx.GetSndAddr(), txMaxTotalCost)
+			if !addressHasEnoughBalance {
+				numTxsWithInitialBalanceConsumed++
+				continue
+			}
+		}
+
+		snapshot := txs.accounts.JournalLen()
+
+		gasConsumedByMiniBlockInReceiverShard := mapGasConsumedByMiniBlockInReceiverShard[receiverShardID]
```

Comment:
you can take this out into a function as well. This function is too big.

Reply:
Done